### ISSUE TITLE
Disabling cache doesn’t work reliably in tests

### ISSUE THREAD

**Issue by nphilipp** [Krukov/cashews#119]
Hi!

We use `cashews` in the upcoming implementation of [Fedora Message Notifications](https://github.com/fedora-infra/fmn) (see the `fmn-next` branch), e.g. to cache requests to backend services which we don’t want to hammer too hard. For testing, we want to disable the cache so every such request would get its own version of (mocked) backend results.

I'm currently working on adding caching to one of these backend services and am struggling because one of the tests gets the cached result of a previously run test (which then fails the test). This is with 5.0.0, I've tested it with 4.7.1 where it works as I expect it. Bisecting the 4.7.1..5.0.0 range, I tracked the change in behavior down to commit 4429f0182b77a42a5f5601137b9ca15fb2281f22 which adds the `@lru_cache()` decorator to `Cache._get_backend_and_config()`. Commenting out that line fixes the issue for me.

Here’s a small script `test_cashews_disabled.py` reproducing the issue:
```
import pytest
from cashews import cache

SIDE_EFFECT = "the side-effect"


@pytest.fixture(autouse=True)
def setup_mem_cache():
    cache.setup("mem://")


@pytest.fixture(autouse=True)
def disable_cache(setup_mem_cache):
    with cache.disabling():
        yield


@pytest.fixture
def change_the_side_effect():
    global SIDE_EFFECT

    SIDE_EFFECT = "the changed side-effect"

    yield


@cache(ttl="1h")
async def function_to_test():
    return SIDE_EFFECT


def test_unrelated():
    assert "BOOP"


@pytest.mark.asyncio
async def test_with_original_side_effect():
    assert await function_to_test() == "the side-effect"


@pytest.mark.asyncio
async def test_with_different_side_effect(change_the_side_effect):
    assert await function_to_test() == "the changed side-effect"
```

To run, e.g. install `cashews`, `pytest` and `pytest-asyncio` into a virtualenv and run `pytest -v test_cashews_disabled.py`.

Here’s the result I get with 5.0.0:
```
(cashews_disabled_test) nils@makake:~/test/python/cashews_disabled_test> pip show cashews
Name: cashews
Version: 5.0.0
Summary: cache tools with async power
Home-page: https://github.com/Krukov/cashews/
Author: Dmitry Kryukov
Author-email: glebov.ru@gmail.com
License: MIT
Location: /home/nils/.virtualenvs/cashews_disabled_test/lib/python3.11/site-packages
Requires: 
Required-by: 
(cashews_disabled_test) nils@makake:~/test/python/cashews_disabled_test> pytest -v test_cashews_disabled.py 
======================================== test session starts =========================================
platform linux -- Python 3.11.1, pytest-7.2.1, pluggy-1.0.0 -- /home/nils/.virtualenvs/cashews_disabled_test/bin/python
cachedir: .pytest_cache
rootdir: /home/nils/test/python/cashews_disabled_test
plugins: asyncio-0.20.3
asyncio: mode=Mode.STRICT
collected 3 items                                                                                    

test_cashews_disabled.py::test_unrelated PASSED                                                [ 33%]
test_cashews_disabled.py::test_with_original_side_effect PASSED                                [ 66%]
test_cashews_disabled.py::test_with_different_side_effect FAILED                               [100%]

============================================== FAILURES ==============================================
__________________________________ test_with_different_side_effect ___________________________________

change_the_side_effect = None

    @pytest.mark.asyncio
    async def test_with_different_side_effect(change_the_side_effect):
>       assert await function_to_test() == "the changed side-effect"
E       AssertionError: assert 'the side-effect' == 'the changed side-effect'
E         - the changed side-effect
E         ?    --------
E         + the side-effect

test_cashews_disabled.py:43: AssertionError
====================================== short test summary info =======================================
FAILED test_cashews_disabled.py::test_with_different_side_effect - AssertionError: assert 'the side-effect' == 'the changed side-effect'
==================================== 1 failed, 2 passed in 0.03s =====================================
(cashews_disabled_test) nils@makake:~/test/python/cashews_disabled_test>
```

And here's the same with 4.7.1:
```
(cashews_disabled_test) nils@makake:~/test/python/cashews_disabled_test> pip show cashews
Name: cashews
Version: 4.7.1
Summary: cache tools with async power
Home-page: https://github.com/Krukov/cashews/
Author: Dmitry Kryukov
Author-email: glebov.ru@gmail.com
License: MIT
Location: /home/nils/.virtualenvs/cashews_disabled_test/lib/python3.11/site-packages
Requires: 
Required-by: 
(cashews_disabled_test) nils@makake:~/test/python/cashews_disabled_test> pytest -v test_cashews_disabled.py 
======================================== test session starts =========================================
platform linux -- Python 3.11.1, pytest-7.2.1, pluggy-1.0.0 -- /home/nils/.virtualenvs/cashews_disabled_test/bin/python
cachedir: .pytest_cache
rootdir: /home/nils/test/python/cashews_disabled_test
plugins: asyncio-0.20.3
asyncio: mode=Mode.STRICT
collected 3 items                                                                                    

test_cashews_disabled.py::test_unrelated PASSED                                                [ 33%]
test_cashews_disabled.py::test_with_original_side_effect PASSED                                [ 66%]
test_cashews_disabled.py::test_with_different_side_effect PASSED                               [100%]

========================================= 3 passed in 0.02s ==========================================
(cashews_disabled_test) nils@makake:~/test/python/cashews_disabled_test>
```

**Comment by Krukov** [comment]
Hi @nphilipp . Thanks for such a detailed report. 

Yes the issue is in `lru_cache`.  It's so dramatic that the cache library has a cache invalidation issue.

I have had doubts when I added it. I will fix the issue soon

**Comment by Krukov** [comment]
Fixed in 5.1.0 and upper. 

Incase of any problem let me know 

Thanks.

**Comment by nphilipp** [comment]
Thanks for the quick fix!
